Fix for --controls option is not working as expected.

Vasu1105 opened the pull request:
Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

<!--- Provide a short summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail, what problems does it solve? -->
As mentioned in issue #5215 the control block also gets evaluated and due to which the `--controls ` option is not working as expected.
The reason behind that the filtering of control here https://github.com/inspec/inspec/blob/master/lib/inspec/profile.rb#L228 is happening after evaluating the controls which are happening here https://github.com/inspec/inspec/blob/master/lib/inspec/profile.rb#L220

which calls this https://github.com/inspec/inspec/blob/2f41c16ad444eaf69f7ec77f4ff681f35aad9210/lib/inspec/profile_context.rb#L154 and then this https://github.com/inspec/inspec/blob/2f41c16ad444eaf69f7ec77f4ff681f35aad9210/lib/inspec/profile_context.rb#L162

which then calls this control definition https://github.com/inspec/inspec/blob/2f41c16ad444eaf69f7ec77f4ff681f35aad9210/lib/inspec/control_eval_context.rb#L54
and describe definition  (descrhttps://github.com/inspec/inspec/blob/2f41c16ad444eaf69f7ec77f4ff681f35aad9210/lib/inspec/control_eval_context.rb#L66

both these methods instantiate Inspec::rule https://github.com/inspec/inspec/blob/2f41c16ad444eaf69f7ec77f4ff681f35aad9210/lib/inspec/rule.rb#L46
which basically causing the skipped control is getting evaluated (note not the describe block)

I have added a method in the control_eval_context.rb to filter the controls before evaluating it. I need to add the filter in describe method also because when we are having only describe block without control block the methods creates auto-generated control as described here https://github.com/inspec/inspec/blob/2f41c16ad444eaf69f7ec77f4ff681f35aad9210/lib/inspec/control_eval_context.rb#L63
so I am making sure those describe blocks are also get filtered when specifying --controls option.

## Related Issue
<!--- If you are suggesting a new feature or change, please create an issue first -->
<!--- Please link to the issue, discourse, or stackoverflow here: -->
Fixes #5215
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the **CONTRIBUTING** document.

Vasu1105 committed:
Added test to catch the control execution issue while using controls option

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

Vasu1105 committed:
Updated test to match the exact string

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

Vasu1105 committed:
Revert "Updated test to match the exact string"

This reverts commit 2048c4dff6689342cd7ef42ea9ba98a4c6715723.

Vasu1105 committed:
Fixed test file for syntax issues

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

Vasu1105 committed:
Fixes #5215 : When we are using --controls option it was also evaluating the control block of not included controls in the controls option as it was filtering the controls from the list after evaluating updated the logic so that it gets evaluated after filtering. Removed filter_controls methods as no more using that

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

clintoncwolfe commented on test/functional/inspec_exec_test.rb:
```diff
@@ -178,6 +178,25 @@ def stderr
   end
 
+  it "executes only specified controls when selecting the controls by literal names" do
+    inspec("exec " + File.join(profile_path, "filter_table") + " --no-create-lockfile --controls foo")
```
Why did you choose the filter_table fixture? I don't think this problem has anything to do with FilterTable. I would suggest either finding a different profile, or creating a new one just for this purpose.

Vasu1105 replied:
Yes after checking the impact decided to remove this from here. I thought since we are using --controls options in test with this profile that will be good place to include. But yes definitely this is not the right place to have this. Making the required changes.

clintoncwolfe commented on test/functional/inspec_exec_test.rb:
```diff
@@ -178,6 +178,25 @@ def stderr
+Version: 0.1.0
+Target:  local://
+
+  ✔  foo: a thing
```
This is likely too brittle of a test - the check marks vary under Windows, for example. I suggest you write tests that look specifically for what you are looking for. For example, 

```
_(out.stdout).wont_include "bar"
_(out.stdout).wont_include "baz"
```

Vasu1105 committed:
Fixing test cases

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

Vasu1105 committed:
Removed readme file from test profile

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

Vasu1105 committed:
Fixes 5215: While using --controls options the control block was also getting evaluated as filtering of the control was happening after evaluating so added the filter logic in the control_eval_context. Also when we have describe block outside control block we are we
generating a control for them automatically and then execute due that also becomes a control and has to add same logic to filter the control in that mehtod

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

Vasu1105 committed:
minor code fix

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

Vasu1105 committed:
Fixed failing test in inputs_test.rb

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

clintoncwolfe commented on test/functional/inspec_exec_test.rb:
```diff
@@ -178,6 +178,27 @@ def stderr
+    assert_exit_code 0, out
+  end
+
+  it "executes only specified controls when selecting the controls by literal names" do
```
This selects by regex, not by literal name

Vasu1105 replied:
Updated.

clintoncwolfe commented on test/fixtures/profiles/controls-option-test/controls/example.rb:
```diff
@@ -0,0 +1,69 @@
+# you can also use plain tests
+describe file('/') do
+  it { should be_directory }
+end
```
This file has a lot of extra material in it that is not being tested. Either it should be removed or it should be tested - fixture files should be minimal. Thanks!

Vasu1105 replied:
Thanks @clintoncwolfe .Did the required changes.

clintoncwolfe reviewed (approved):
Just a couple of minor notes about the test fixture file, nit-picking really. This works really well. I think users will be very happy with this improvement!

Vasu1105 committed:
Updated test for --controls option

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

Vasu1105 committed:
Added code comments and rename control_list_exist? method to controls_list_exist?

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>

Vasu1105 committed:
Added specific regex as for describe block we genrate random control with some random hex which might contain the number that I have used in the rgex due to which test is failing

Signed-off-by: Vasu1105 <vasundhara.jagdale@chef.io>
